src/gui/plugins/visualize_lidar/VisualizeLidar.cc

```diff
@@ -0,0 +1,445 @@
+            << std::endl;
+
+    ignition::gui::App()->findChild<
+        ignition::gui::MainWindow *>()->removeEventFilter(this);
+    return;
+  }
+  if (this->dataPtr->lidar)
+  {
+    root->AddChild(this->dataPtr->lidar);
+    this->dataPtr->initialized = true;
+  }
+}
+
+/////////////////////////////////////////////////
+void VisualizeLidar::LoadConfig(const tinyxml2::XMLElement *)
+{
+  if (this->title.empty())
+    this->title = "Visualize lidar";
+
+  ignition::gui::App()->findChild<
+    ignition::gui::MainWindow *>()->installEventFilter(this);
+}
+
+/////////////////////////////////////////////////
+bool VisualizeLidar::eventFilter(QObject *_obj, QEvent *_event)
+{
+  if (_event->type() == ignition::gazebo::gui::events::Render::kType)
+  {
+    // This event is called in Scene3d's RenderThread, so it's safe to make
+    // rendering calls here
+
+    std::lock_guard<std::mutex> lock(this->dataPtr->serviceMutex);
+    if (!this->dataPtr->initialized)
+    {
+      this->LoadLidar();
+    }
+
+    if (this->dataPtr->lidar)
+    {
+      if (this->dataPtr->visualDirty)
+      {
+        this->dataPtr->lidar->SetWorldPose(this->dataPtr->lidarPose);
+        this->dataPtr->lidar->Update();
+        this->dataPtr->visualDirty = false;
+      }
+    }
+    else
+    {
+      ignerr << "Lidar pointer is not set" << std::endl;
+    }
+  }
+
+  // Standard event processing
+  return QObject::eventFilter(_obj, _event);
+}
+
+//////////////////////////////////////////////////
+void VisualizeLidar::Update(const UpdateInfo &,
+    EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("VisualizeLidar::Update");
+
+  std::lock_guard<std::mutex> lock(this->dataPtr->serviceMutex);
+
+  if (this->dataPtr->lidarEntityDirty)
+  {
+    auto lidarURIVec = common::split(common::trimmed(
+                  this->dataPtr->lidarString), "::");
+    if (lidarURIVec.size() > 0)
+    {
+      auto baseEntity = _ecm.EntityByComponents(
+          components::Name(lidarURIVec[0]));
+      if (!baseEntity)
+      {
+        ignerr << "Error entity " << lidarURIVec[0]
+            << " doesn't exist and cannot be used to set lidar visual pose"
+            << std::endl;
+        return;
+      }
+      else
+      {
+        auto parent = baseEntity;
+        bool success = false;
+        for (auto i = 0ul; i < lidarURIVec.size()-1; i++)
+        {
+          auto children = _ecm.EntitiesByComponents(
+                            components::ParentEntity(parent));
+          bool foundChild = false;
+          for (auto child : children)
+          {
+            std::string nextstring = lidarURIVec[i+1];
+            std::string childname = std::string(
+                            _ecm.Component<components::Name>(child)->Data());
+            if (nextstring.compare(childname) == 0)
+            {
+              parent = child;
+              foundChild = true;
+              if (i+1 == lidarURIVec.size()-1)
```

Comment:
I think the logic in this block assumes the model > link > sensor hierarchy? 

We also support nested models, i.e.g model > model > link > sensor, in which case the logic above may not work?